subsys/net/ip/tcp.c

```diff
@@ -696,18 +696,29 @@ int net_tcp_send_pkt(struct net_pkt *pkt)
 {
 	struct net_context *ctx = net_pkt_context(pkt);
 	struct net_tcp_hdr *tcphdr = NET_TCP_HDR(pkt);
+	bool calc_chksum = false;
 
-	sys_put_be32(ctx->tcp->send_ack, tcphdr->ack);
+	if (sys_get_be32(tcphdr->ack) != ctx->tcp->send_ack) {
+		sys_put_be32(ctx->tcp->send_ack, tcphdr->ack);
+		calc_chksum = true;
+	}
 
 	/* The data stream code always sets this flag, because
 	 * existing stacks (Linux, anyway) seem to ignore data packets
 	 * without a valid-but-already-transmitted ACK.  But set it
 	 * anyway if we know we need it just to sanify edge cases.
 	 */
-	if (ctx->tcp->sent_ack != ctx->tcp->send_ack) {
+	if (ctx->tcp->sent_ack != ctx->tcp->send_ack &&
+		(tcphdr->flags & NET_TCP_ACK) == 0) {
+		calc_chksum = true;
 		tcphdr->flags |= NET_TCP_ACK;
 	}
 
+	if (calc_chksum) {
+		NET_TCP_HDR(pkt)->chksum = 0;
```

Comment:
The NET_TCP_HDR() macro was removed by #691 so this part needs changing.

You can call

net_tcp_set_chksum(pkt, pkt->frags);

to calculate the checksum.